File: game-logic-core/src/main/scala/fwc/game/board/MilitaryUnit.scala
package fwc.game.board

import fwc.{JsonParsable, JsonSerializable}
import fwc.game.houses.HouseType
import scala.util.Try
import ujson.Value

case class MilitaryUnit(
                         house: HouseType,
                         unitType: MilitaryUnitType,
                         isDefeated: Boolean = false,
                         garrisonDefensePoints: Int = 0
                       ) extends JsonSerializable {

  def toJson: ujson.Value = {
    val milUnit = ujson.Obj(
      "house" -> ujson.Str(house.toString),
      "type" -> ujson.Str(unitType.toString)
    )
    if isDefeated then milUnit.obj addOne ("isDefeated" -> ujson.Bool(isDefeated))
    if unitType == MilitaryUnitGarrison then milUnit.obj addOne ("defPoints" -> ujson.Num(garrisonDefensePoints))
    milUnit
  }

}

object MilitaryUnit extends JsonParsable {
  override def fromJson(json: Value): MilitaryUnit = {
    MilitaryUnit(
      HouseType.fromString(json.obj("house").str),
      MilitaryUnitType.fromString(json.obj("type").str),
      Try[Boolean](json.obj("isDefeated").bool).getOrElse(false),
      Try[Double](json.obj("defPoints").num).getOrElse(.0).toInt
    )
  }
}

File: bots/server/game_state/military_unit.py
from server.game_state.house_type import HouseType
from server.game_state.military_unit_type import MilitaryUnitType


class MilitaryUnit:
    def __init__(self,
                 house: HouseType,
                 unit_type: MilitaryUnitType,
                 is_defeated: bool = False,
                 garrison_defence_points: int = 0
                 ):
        self.house = house
        self.unit_type = unit_type
        self.is_defeated = is_defeated
        self.garrison_defence_points = garrison_defence_points

    @classmethod
    def from_json(cls, json):
        return cls(
            HouseType[json['house'].upper()],
            MilitaryUnitType.from_str(json['type']),
            json['isDefeated'] if 'isDefeated' in json else False,
            json['defPoints'] if 'defPoints' in json else 0
        )